File: src/main/java/net/vidageek/games/regex/task/MatcherTargets.java
package net.vidageek.games.regex.task;

import java.util.Arrays;
import java.util.Collections;
import java.util.Iterator;
import java.util.List;

import com.google.common.base.Joiner;
import com.google.common.collect.Lists;

public class MatcherTargets implements Iterable<String> {

	private final Escaper escape = new Escaper();
	private final List<String> matcherTargets = Lists.newArrayList();

	private MatcherTargets(final String... othersMatchersTargets) {
		matcherTargets.addAll(Arrays.asList(othersMatchersTargets));
	}

	public static MatcherTargets from(final String... matchingTargets) {
		return new MatcherTargets(matchingTargets);
	}

	private List<String> scapeTarges() {
		return escape.applyAll(matcherTargets);
	}

	public Iterator<String> iterator() {
		return Collections.unmodifiableCollection(matcherTargets).iterator();
	}

	public String asHtml() {
		return swapLastComma("<code>" + Joiner.on("</code>, <code>").join(scapeTarges()) + "</code>");
	}

	private String swapLastComma(final String string) {
		for (int i = string.length() - 1; i >= 0; i--) {
			if (string.charAt(i) == ',') {
				return string.substring(0, i) + " e" + string.substring(i + 1);
			}
		}
		return string;
	}

}

File: src/main/scala/net/vidageek/games/regex/task/MatcherTargets.scala
package net.vidageek.games.regex.task

import com.google.common.base.Joiner
import scala.collection.JavaConverters._

class MatcherTargets private(matcherTargets: List[String]) extends Iterable[String] {
  private val escape = new Escaper()
  private def scapeTarges() = escape.applyAll(matcherTargets asJava)

  def iterator(): Iterator[String] = matcherTargets.iterator

  def asHtml(): String = swapLastComma("<code>" + Joiner.on("</code>, <code>").join(scapeTarges()) + "</code>")

  private def swapLastComma(string: String) = {
    if (string contains ",")
      string.reverse.replaceFirst(",", "e ").reverse
    else 
      string
  }
}

object MatcherTargets {
  def from(matchingTargets: String*): MatcherTargets = new MatcherTargets(matchingTargets.toList)
  def from(matchingTargets: Array[String]): MatcherTargets = new MatcherTargets(matchingTargets.toList)
}
